src/workerd/api/r2-multipart.h

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2017-2022 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
+#pragma once
+
+#include "r2-bucket.h"
+
+#include <workerd/jsg/jsg.h>
+
+namespace workerd::api::public_beta {
+class R2MultipartUpload: public jsg::Object {
+  public:
+    R2MultipartUpload(kj::String key, kj::String uploadId, jsg::Ref<R2Bucket> bucket):
+      key(kj::mv(key)), uploadId(kj::mv(uploadId)), bucket(kj::mv(bucket)) {}
+
+    kj::String getKey() const { return kj::str(key); }
+    kj::String getUploadId() const { return kj::str(uploadId); }
```

Comment:
Trying to remember off the top of my head but jsg supports it, but it would be better if these returned `kj::StringPtr` to avoid extra allocations and copies.

Reply:
(not critical, of course, especially since these are lazy data properties)

Reply:
Seems to work, thanks for these tips! Should give the KJ tour a better read.

Reply:
I suppose I could also optimize some stuff by using kj::StringPtr on some internal kj::str copies I'm currently doing, e.g. in code like this:

```
return bucket.get()->uploadPart(
  js, kj::str(key), kj::str(uploadId), partNumber, kj::mv(value), errorType
);
```

Here it's probably still quite simple to reason about why that's correct, since the multipart object lives longer than the uploadPart call & the uploadPart call doesn't modify its arguments, but I don't know if there's enough performance benefit to bother reasoning about this & opening myself up to memory bugs in case the called function does want to modify its arguments. I miss having a type system that would let me enforce immutable references :disappointed:

Reply:
Although, i'm not sure if kj::StringPtr would be correct for that example. That key makes it into a `uploadPartBuilder.setObject(key)` which then makes its way into a kj::Promise associated with a http put request, and that makes reasoning about lifetimes a bit more difficult I think. Probably still correct, but who knows :shrug: